### PR TITLE
Improve error reporting for incompatible lock files

### PR DESCRIPTION
**Pull request by tgoyne**
Adds a better error message that tries to explain what's wrong.

@alazier @jpsim

### CODE REVIEW

File: Realm/RLMRealm.mm [hunk 1]
```diff
@@ -302,20 +302,32 @@ - (instancetype)initWithPath:(NSString *)path key:(NSData *)key readOnly:(BOOL)r
...
             error = make_realm_error(RLMErrorFileExists, ex);
         }
-        catch (File::AccessError &ex) {
+        catch (File::AccessError const& ex) {
             error = make_realm_error(RLMErrorFileAccessError, ex);
         }
-        catch (exception &ex) {
+        catch (IncompatibleLockFile const&) {
+            NSString *err = @"Realm file is currently open in another process "
+                             "which cannot shared access with this process. All "
```
Comment (jpsim) [inline Realm/RLMRealm.mm]: s/shared/share/

**Comment by jpsim** [comment]
Should we be displaying an error alert in the browser as well?

**Comment by tgoyne** [comment]
Browser really should be displaying all errors when opening realms...

Should be easy to add.

**Comment by jpsim** [comment]
Tracking that in #1354.

**Comment by jpsim** [comment]
:+1:

**Comment by alazier** [comment]
Can we add a test for this?

**Comment by tgoyne** [comment]
No, it requires multiple processes of different architectures.

**Comment by alazier** [comment]
:+1:

**Comment by alazier** [comment]
We could generate lock files from another architecture, and then spawn a process which opens these files. Then it would appear that we are accessing lock files created from another architecture.

**Comment by tgoyne** [comment]
Added a test which mangles the lock file directly.

File: CHANGELOG.md [hunk 2]
```diff
@@ -27,6 +27,8 @@ x.x.x Release notes (yyyy-MM-dd)
 * Fixed a crash when adding primary keys to older realm files with no primary
   keys on any objects.
 * Fixed a crash when removing a primary key in a migration.
+* Improve the erorr message when a Realm file is opened in multiple processes
```
Comment (jpsim) [inline CHANGELOG.md]: error

**Comment by tgoyne** [comment]
Result of opening a file in the browser that's already open in a 32-bit sim now that the browser actually displays errors:

image

**Pull request merged**